common/captcha/CaptchaSolverInterface.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC, abstractmethod
+import os
+import cv2
+import time
+
+
+class SolvedCaptcha:
+    """
+    A class interface for representing solved Captchas.
+    """
+
+    def __init__(self, image, text, answer=None, out_dir=None):
+        """
+        :param image: Unprocessed Captcha image
+        :param text: The text contained in the captcha image. (eg: 12+3=)
+        :param answer: The solved captcha's answer (eg: 15). In the 'jumbled text' captchas this is the same as 'text'.
+        :param out_dir: Where to write captchas to upon notify_correct() or notify_incorrect()
+        """
+        self.image = image
+        self.text = text
+        self.answer = answer
+        self.out_dir = out_dir
+
+    def notify_correct(self):
```

Comment:
It seems like the `notify_*` functions really just store the captcha, yeah? It's not clear to me what they "notify".

Is there ever a scenario in which we would want `notify_*` to do more? If not, would it make more sense to call these `store_*` to be a bit more precise?

Another option would be to expose a singular `dump` method in our interface that just writes an implementation-defined stream of bytes useful when debugging to a caller-specified file or something. As it stands, the caller has to pass in an `out_dir` into the solver already, so the callers are already somewhat responsible for managing this. This would simplify our API surface without materially damaging the utility of the solver, IMO. Thoughts?

Reply:
I've renamed it to `save_captcha` with a parameter 'correct' which determines whether it goes in the 'correct' or 'incorrect' folder.

As for the dump method, I'm not really sure how that would be implemented. If the user wanted to access the raw captcha image they could simply get the `SolvedCaptcha.image` object variable. 

In my example, this would be an OpenCV image, which is stored in a NumPy Ndarray. There are various ways to dump that to byte streams like `numpy.ndarray.tobytes`. 

However, I would advise against enforcing the types held in SolvedCaptcha.image, as the format/structure this will be in is dependent on the library choice in the Captcha Solver.

Reply:
Yeah, this seems good. I didn't really understand the "this is also training data" angle before, so this all makes more sense to me now.